src/System.Management.Automation/engine/runtime/Profiler.cs

```diff
@@ -0,0 +1,402 @@
+            if (IsEnabled())
+            {
+                unsafe
+                {
+                    EventData* eventPayload = stackalloc EventData[2];
+
+                    eventPayload[0] = new EventData
+                    {
+                        Size = sizeof(Guid),
+                        DataPointer = ((IntPtr)(&ScriptBlockId))
+                    };
+                    eventPayload[1] = new EventData
+                    {
+                        Size = sizeof(int),
+                        DataPointer = ((IntPtr)(&SequencePointPosition))
+                    };
+
+                    WriteEventCore(eventId: 1, eventDataCount: 2, eventPayload);
+                }
+            }
+        }
+
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            base.OnEventCommand(command);
+
+            if (command.Command == EventCommand.Disable)
+            {
+                // At the end of the profile session, we send all the metadata.
+                ProfilerRundownEventSource.LogInstance.WriteRundownEvents();
+            }
+        }
+    }
+
+    /// <summary>
+    /// This class is a source of rundown (meta data) events for script block profiling.
+    /// Guid is {f348266e-dde6-5590-4bc9-10e1e2b6fe16}.
+    /// </summary>
+    [EventSource(Name = "Microsoft-PowerShell-Profiler-Rundown")]
+    internal class ProfilerRundownEventSource : EventSource
+    {
+        internal static ProfilerRundownEventSource LogInstance = new ProfilerRundownEventSource();
+
+        [Event(2)]
+        public void ScriptBlockRundown(
+            Guid ScriptBlockId,
+            string ScriptBlockText)
+        {
+            // It is not performance critical so we use the standard method overload.
+            WriteEvent(
+                eventId: 2,
+                ScriptBlockId,
+                ScriptBlockText);
+        }
+
+        [Event(3)]
+        public void SequencePointRundown(
+            Guid ScriptBlockId,
+            int SequencePointCount,
+            int SequencePoint,
+            string File,
+            int StartLineNumber,
+            int StartColumnNumber,
+            int EndLineNumber,
+            int EndColumnNumber,
+            string Text,
+            int StartOffset,
+            int EndOffset)
+        {
+            // It is not performance critical so we use the standard method overload.
+            WriteEvent(
+                eventId: 3,
+                ScriptBlockId,
+                SequencePointCount,
+                SequencePoint,
+                File,
+                StartLineNumber,
+                StartColumnNumber,
+                EndLineNumber,
+                EndColumnNumber,
+                Text,
+                StartOffset,
+                EndOffset);
+        }
+
+        [NonEvent]
+        internal void WriteRundownEvents()
+        {
+            foreach (var pair in CompiledScriptBlockData.GetCompiledScriptBlockTable())
```

Comment:
This isn't ideal as it may write out much more than necessary, but it may also miss some script blocks that are hit while profiling.

The ConditionalWeakTable does not prevent garbage collection, so if, e.g. we cleared the script block cache while profiling, scripts that had exited while profiling would be missing the rundown data.

Reply:
- Yes, it is not ideal. For the simple test I shared I see a size of the table is 86. In big application (Pester) it could be 500 or 1000 - it does not look so critical bad.
- I tried ConcurrentDictionary instead of ConditionalWeakTable. You can revert last commit and see the case. My idea was to register scriptblock in Compile() method so that we was independent from scriptblock cache. It is more predictable but we need to clean the buffer somehow.
    - for ConcurrentDictionary  case I cleaned up the buffer at EventSource disable time. It is not ideal too. It duplicates the scriptblock cache.
    - for ConditionalWeakTable we could temporary stop to clean up the scriptblock cache until we disable EventSource.

Reply:
The profiler could take a snapshot of the script block cache when starting. This snapshot would hold references so it wouldn't matter if the script block cache was cleared.

Then, script block creation would also add to that snapshot.

This way, the memory is owned by the profiler and can be cleared when profiling is disabled allowing garbage collection to do it's thing.

Reply:
The snapshot does not help in a scenario with long live scriptblocks. Such scriptblocks could be tracked with ConditionalWeakTable. We could avoid ConditionalWeakTable if there is a way to identify running scripts.

Reply:
I think you could gather by breaking every runspace (I think the debugger can do this) and walking the stack of each runspace.

But I think it would be simpler to just take a snapshot of the scriptblock cache. IIRC, the only scriptblock that doesn't end up in the cache comes from the command line.

Regardless, ConditionalWeakTable is likely not what you want because you don't control keeping the data you need alive.

Reply:
I was playing around with it this weekend, and ConditionWeakTable doesn't make sense unless we tie each cache entry to an InternalProfiler. Best I could come up with is https://github.com/IISResetMe/PowerShell/commit/a33083f85bb968e8d9311b31c7d6e847bc23b3b4, adds tracking of RunspaceId to sequence points and limits the lifetime of the compiled script block cache

Reply:
@IISResetMe Thanks! I was thinking about RunspaceId too. I think it is useful to add this in perf events.
As for cache I believe we could use ConcurrentDictionary instead of a custom implementation.

Reply:
The motivation for doing an ordered set-like cache like this is to preserve insertion order - but for the `InternalProfiler` use case that isn't strictly necessary, `ConcurrentDictionary<CompiledScriptBlockData,byte>` or similar would work just fine. 

One thing I'm still keen on trying to figure out is if we can somehow get the compiler to tag all position updates in nested scriptblock compilations with the ID of the immediate parent (or root) block (or alternatively the id of the dispatching runspace), so that we can re-construct the call tree:

```
>> Measure-Script { 1..3 |ForEach-Object -Parallel { $_ } }

ParentId  ScriptBlockId  RunspaceId  ScriptText
       0              1           1  { 1..3 |ForEach-Object -Parallel { $_ } }
       1              2           2  { $_ }
       1              3           3  { $_ }
       1              4           4  { $_ }
```

This should make it possible to run multiple profilers in parallel while still allowing them to individually filter events related to themselves

Reply:
Now I replace ConditionalWeakTable  with ConcurrentDictionary, add an initialization from scriptblock cache, simplify the scriptblock cache ( we cached ScriptBlock but never use it directly - clone before use but Clone() method actually
copies a reference to CompiledScriptBlock so we can cache directly  CompiledScriptBlock)ю

Reply:
@lzybkr Do you have any ideas for how to track the "origin" or "root dispatcher" for the individual script blocks as described above? It would make it so much easier to consume the events in any tool/context.

Reply:
Jason said Debugger has a callstack. I am looking...

Reply:
I added Runspace and parent scriptblock Ids in perf event. I am still not sure it works as expected - need a review.
Also I added a typed output and set Start Opcode on perf event - need to find how to issue Stop event.

Reply:
I pulled some commits and it's ready for further discussion. The PR description has also been updated.